### PR TITLE
Passing labels when calling pd.DataFrame predict_proba and classificiation

### LINKED ISSUE #579: [BUG] GroupedPredictor inconsistency for predict_proba having different classes per group
Hello scikit-lego users.
While using predict_proba paired with GroupedPredictor and a classifier on a df with different labels per group, the final matrix is collapsed to the left, without caring about the label order.
This yields inconsistencies in the final output, especially while using an high number of labels.
To achieve a sound result every label should appear at least once in every group, that is somehow unrealistic.

Here a snippet of code:
``` python
import pandas as pd
import numpy as np

from sklego.meta import GroupedPredictor
from sklearn.linear_model import LogisticRegression


np.random.seed(43)

group_size = 5
n_groups = 2
df = pd.DataFrame({
    "group": ["A"] * group_size + ["B"] * group_size,
    "x": np.random.normal(size=group_size * n_groups),
    "y": np.hstack([
        np.random.choice([0, 1, 2], size=group_size),
        np.random.choice([0, 2], size=group_size),
        ])
})

print(df.groupby('group').agg({'y': set}))


X, y = df[["x", "group"]], df["y"]
model = GroupedPredictor(LogisticRegression(), groups=["group"])
_ = model.fit(X, y)
y_prob = model.predict_proba(X)

print(y_prob.round(2))

```

Outputs:
``` python

>>>                y
>>> group           
>>> A      {0, 1, 2}
>>> B         {0, 2}

>>> [[0.45 0.19 0.36]#grp A
>>> [0.3  0.23 0.47]
>>> [0.37 0.21 0.42]
>>> [0.35 0.22 0.44]
>>> [0.53 0.16 0.31]
>>> [0.79 0.21  nan]#grp B
>>> [0.8  0.2   nan]
>>> [0.81 0.19  nan]
>>> [0.81 0.19  nan]
>>> [0.79 0.21  nan]]

# Expected:
>>> [[0.45 0.19 0.36]#grp A
>>> [0.3  0.23 0.47]
>>> [0.37 0.21 0.42]
>>> [0.35 0.22 0.44]
>>> [0.53 0.16 0.31]
>>> [0.79 nan  0.21]#grp B
>>> [0.8  nan   0.2]
>>> [0.81 nan  0.19]
>>> [0.81 nan  0.19]
>>> [0.79 nan  0.21]]

```

### PR DESCRIPTION
**Pull request by fabioscantamburlo**
# Description

When calling **__predict_single_group** method with argument _method_ == _'predict_proba'_ ensure that the model is a classifier.
If yes,  pass as _extra_kwargs_ to **pd.DataFrame** columns = labels of the model. 

This is a non-breaking change that solves an inconsistency using **predict_proba** with sub groups with different labels.

Fixes # [[BUG] GroupedPredictor inconsistency for predict_proba having different classes per group](https://github.com/koaning/scikit-lego/issues/579)

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [ ] My code follows the style guidelines (flake8)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (also to the readme.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added tests to check whether the new feature adheres to the sklearn convention
- [x] New and existing unit tests pass locally with my changes

Fixes #579

### CODE REVIEW

File: tests/test_meta/test_grouped_predictor.py [hunk 1]
```diff
@@ -72,6 +108,44 @@ def test_chickweight_can_do_fallback_proba():
     assert mod.predict_proba(to_predict).shape == (2, 2)
     assert (mod.predict_proba(to_predict)[0] == mod.predict_proba(to_predict)[1]).all()
 
+@pytest.mark.parametrize("df_predict_proba_request", [
```
Comment (fabioscantamburlo) [inline tests/test_meta/test_grouped_predictor.py:111]: Test generic boundary conditions

File: tests/test_meta/test_grouped_predictor.py [hunk 2]
```diff
@@ -72,6 +108,44 @@ def test_chickweight_can_do_fallback_proba():
...
+def test_predict_proba_has_same_columns_as_distinct_labels(df_predict_proba_request):
+    mod = GroupedPredictor(estimator=LogisticRegression(), groups='group')
+    X, y = df_predict_proba_request[["group", "x"]], df_predict_proba_request["y"]
+    _ = mod.fit(X, y)
+    y_proba = mod.predict_proba(X)
+
+    # Ensure the number of col output is always equal to the cardinality of the labels
+    assert(len(df_predict_proba_request['y'].unique()) == y_proba.shape[1])
+
+def test_predict_proba_correct_nulls_different_labels(df_predict_proba):
```
Comment (fabioscantamburlo) [inline tests/test_meta/test_grouped_predictor.py:126]: Test specific situation

**Comment by fabioscantamburlo** [comment]
Fixes koaning-scikit-lego#579

File: sklego/meta/grouped_predictor.py [hunk 3]
```diff
@@ -292,9 +294,15 @@ def __predict_single_group(self, group, X, method="predict"):
                     f"Found new group {group} during predict with use_global_model = False"
                 )
 
+        # Ensure to mantain label order if it's called with predict_proba
+        if is_classifier(group_predictor) and method == "predict_proba":
+            extra_kwargs = {"columns": group_predictor.classes_}
+
```
Comment (FBruzzesi) [inline sklego/meta/grouped_predictor.py:300]: Couple of nitpicks:
1. The comment may be misleading, the order is actually maintained by the behaviour of [pd.concat](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) (with `axis=0`), here we just register the column names (labels) that appear in the specific group
2. Probably personal preference but I would opt for something along the line:
    ```py
    if condition:
        extra_kwargs = {"columns": group_predictor.classes_}
    else:
        extra_kwargs = {}
    ```
    where you don't initialize the extra_kwargs at the beginning of the function

File: tests/test_meta/test_grouped_predictor.py [hunk 4]
```diff
@@ -13,6 +13,42 @@
 from tests.conftest import general_checks, select_tests
 
 
+@pytest.fixture(name="df_predict_proba")
+def get_df_different_classes_classification():
```
Comment (FBruzzesi) [inline tests/test_meta/test_grouped_predictor.py:17]: Isn't this a specific instance/case of `df_predict_proba_request` fixture?

File: tests/test_meta/test_grouped_predictor.py [hunk 5]
```diff
@@ -13,6 +13,42 @@
...
+        [
+            np.random.choice([0, 1, 2], size=group_size),
+            np.random.choice([0, 2], size=group_size),
+        ]
+    )
+    df = pd.DataFrame({"group": group_col, "x": x_col, "y": y_col})
+    return df
+
+@pytest.fixture(name="df_predict_proba_request")
+def my_custom_df(request):
```
Comment (FBruzzesi) [inline tests/test_meta/test_grouped_predictor.py:35]: - Following the repo convention you can name the fixture directly instead of using the `name` keyword. 
- If possible I would follow some naming convention as in [conftest.py](https://github.com/koaning/scikit-lego/blob/main/tests/conftest.py). Maybe something along the line of `random_xy_grouped_clf_different_classes` ?(holy moly naming is hard 😂)

**Review by FBruzzesi — COMMENTED** [review]
Additionally to the comments in the code block, I would argue that `__predict_groups` should fill the missing labels from groups with **zeros** and **do not** return `NaN`s.

This adjustment would require some modification in the tests as well

**Comment by fabioscantamburlo** [comment]
#579

File: tests/test_meta/test_grouped_predictor.py (hunk 5 above)
Comment (fabioscantamburlo) [reply tests/test_meta/test_grouped_predictor.py:35]: Yes you're right. 
Changing it to follow the repo's convention. !

**Review by FBruzzesi — COMMENTED** [review]
Tests feedback

File: tests/test_meta/test_grouped_predictor.py [hunk 6]
```diff
@@ -108,43 +91,93 @@ def test_chickweight_can_do_fallback_proba():
...
+    df_proba = pd.concat(
+        [random_xy_grouped_clf_different_classes["group"], pd.DataFrame(y_proba)],
+        axis=1,
+    )
+
+    # Ensure for the common labels there are no zeros
+    in_common_labels = [0, 2]
+    for _, grp in df_proba.groupby("group"):
+        for label in in_common_labels:
+            assert (grp.loc[:, label] != 0).all()
```
[the remark this answers is not in the document]
Comment (fabioscantamburlo) [reply tests/test_meta/test_grouped_predictor.py:180]: Yep, I agree with you.
Let's try to make it more general 👍

**Review by FBruzzesi — APPROVED** [review]
This is great 👌 Thanks so much
